src/ModelContextProtocol.Core/Authentication/GenericOAuthProvider.cs

```diff
@@ -0,0 +1,544 @@
+/// </summary>
+/// <param name="authorizationUrl">The authorization URL that the user needs to visit.</param>
+/// <param name="redirectUri">The redirect URI where the authorization code will be sent.</param>
+/// <param name="cancellationToken">The cancellation token.</param>
+/// <returns>A task that represents the asynchronous operation. The task result contains the authorization code if successful, or null if the operation failed or was cancelled.</returns>
+/// <remarks>
+/// <para>
+/// This delegate provides SDK consumers with full control over how the OAuth authorization flow is handled.
+/// Implementers can choose to:
+/// </para>
+/// <list type="bullet">
+/// <item><description>Start a local HTTP server and open a browser (default behavior)</description></item>
+/// <item><description>Display the authorization URL to the user for manual handling</description></item>
+/// <item><description>Integrate with a custom UI or authentication flow</description></item>
+/// <item><description>Use a different redirect mechanism altogether</description></item>
+/// </list>
+/// <para>
+/// The implementation should handle user interaction to visit the authorization URL and extract
+/// the authorization code from the callback. The authorization code is typically provided as
+/// a query parameter in the redirect URI callback.
+/// </para>
+/// </remarks>
+public delegate Task<string?> AuthorizationUrlHandler(Uri authorizationUrl, Uri redirectUri, CancellationToken cancellationToken);
+
+/// <summary>
+/// A generic implementation of an OAuth authorization provider for MCP. This does not do any advanced token
+/// protection or caching - it acquires a token and server metadata and holds it in memory. 
+/// This is suitable for demonstration and development purposes.
+/// </summary>
+public class GenericOAuthProvider : IMcpCredentialProvider
+{
+    /// <summary>
+    /// The Bearer authentication scheme.
+    /// </summary>
+    private const string BearerScheme = "Bearer";
+    private readonly Uri _serverUrl;
+    private readonly Uri _redirectUri;
+    private readonly List<string> _scopes;
+    private readonly string _clientId;
+    private readonly string _clientSecret;
+    private readonly HttpClient _httpClient;    private readonly AuthorizationHelpers _authorizationHelpers;
+    private readonly ILogger _logger;
+    private readonly Func<IReadOnlyList<Uri>, Uri?> _authServerSelector;
+    private readonly AuthorizationUrlHandler _authorizationUrlHandler;
+    
+    // Lazy-initialized shared HttpClient for when no client is provided
+    private static readonly Lazy<HttpClient> _defaultHttpClient = new(() => new HttpClient());
+
+    private static readonly JsonSerializerOptions _jsonOptions = new() { PropertyNameCaseInsensitive = true };
+    private TokenContainer? _token;
+    private AuthorizationServerMetadata? _authServerMetadata;    /// <summary>
```

Comment:
```suggestion
    private AuthorizationServerMetadata? _authServerMetadata;

    /// <summary>
```

There's a bunch of other similar line ending issues in this file. I've also noticed trailing whitespace in a lot of the other new files as well. We haven't been the best at avoiding these so far, but it might be a good idea to run the _new_ files through something like dotnet format.